### PR TITLE
Added new stored procedures: TRURSwapTables.java, TRUPSwapTablesSP.java,

### PR DESCRIPTION
**Pull request by johnTcrawford**
TRUPSwapTablesMP.java, which perform a SWAP TABLES (in a replicated, partitioned SP, and partitioned MP context, respectively). Also added these to ddl-nocat.sql, along with definitions of related tables SWAPR & SWAPP (& corrected partitioning of TRUPTableInsert, to use the correct table - TRUP, not BIGP).
TruncateTableLoader.java: in the run() method, added 2 Swap Tables calls, which may or may not occur (randomly), before and after the Truncate; added a swapRatio arg to the constructor, which is used to determine how often those Swap Tables calls occur; added an isStatusSuccess method, to simplify old and new repetitive code.
Benchmark.java: added new swapratio option definition, and added it to arg list in call of new TruncateTableLoader(...).
README: just corrected spelling mistakes.

### CODE REVIEW

File: tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java [hunk 1]
```diff
@@ -255,8 +277,13 @@ public void run() {
                 // just need to fall through and get out
                 throw new RuntimeException(e);
             }
+            // TODO: this seems like useful info, for now, but it could be
+            // removed if it produces too much output
+            log.info("table: " + tableName + " rows sent: " + insertsTried + " inserted: "
+                    + rowsLoaded + " truncates: " + nTruncates + " swaps: " + nSwaps);
```
Comment (prosegay) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java]: make this debug level

File: tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/procedures/TRUPSwapTablesMP.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        row = data.fetchRow(0);
+        long swapScanCountAfter = row.getLong(0);
+        if (truScanCountBefore != swapScanCountAfter || swapScanCountBefore != truScanCountAfter) {
+            throw new VoltAbortException("after swap (scan) counts (" + truScanCountAfter + "," + swapScanCountAfter
+                    + ") do not match reverse of those before (" + truScanCountBefore + "," + swapScanCountBefore + ")");
+        }
+
+        if (shouldRollback != 0) {
+            throw new VoltAbortException("EXPECTED ROLLBACK");
+        }
```
Comment (prosegay) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/procedures/TRUPSwapTablesMP.java]: there's no check in this test that the rollback did what was expected of it is there? that would be a good addition too.

Comment (prosegay) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/procedures/TRUPSwapTablesMP.java]: this comment applies overall, not to just this particular procedure

File: tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java [hunk 3]
```diff
@@ -163,26 +173,46 @@ public void run() {
...
+                if (r.nextInt(100) < swapRatio * 100.) {
+                    shouldRollback = (byte) (r.nextInt(10) == 0 ? 1 : 0);
+                    p = Math.abs(r.nextLong());
+                    clientResponse = client.callProcedure(tableName.toUpperCase() + sp, p, shouldRollback);
+                    if (isStatusSuccess(clientResponse, shouldRollback, "swap", tableName)) {
+                        Benchmark.txnCount.incrementAndGet();
+                        nSwaps++;
+                    }
+                }
+
```
Comment (prosegay) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java]: this is a minor point, I'm not sure what the added complexity of perhaps swapping before or after gets us over just loading, swapping, truncating (with rollbacks at any of those), which is the common use pattern.

Comment (prosegay) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java]: Not so minor, tho, there's no active check of rollback behavior; that would be nice to have.

File: tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/procedures/TRURSwapTables.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            throw new VoltAbortException("after swap (scan) counts (" + truScanCountAfter + "," + swapScanCountAfter
+                    + ") do not match reverse of those before (" + truScanCountBefore + "," + swapScanCountBefore + ")");
+        }
+
+        if (shouldRollback != 0) {
+            throw new VoltAbortException("EXPECTED ROLLBACK");
+        }
+        return results;
+    }
+}
```
Comment (prosegay) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/procedures/TRURSwapTables.java]: minor but would have been really nice to use a common proc to house the code that checks the counts so it only appears once.

File: tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java (hunk 1 above)
Comment (johnTcrawford) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java]: OK, done.

File: tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java (hunk 3 above)
Comment (johnTcrawford) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/TruncateTableLoader.java]: By swapping sometimes before, and sometimes after, truncating you end up sometimes swapping two empty tables, sometimes two full tables, and sometimes one empty with one full (in either direction); if you always swapped before, you would always be swapping two full tables (except the very first swap); if you always swapped after, one of them would always be empty. This is by no means obvious, so I'll add a comment to that effect.

File: tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/procedures/TRURSwapTables.java (hunk 4 above)
Comment (johnTcrawford) [inline tests/test_apps/txnid-selfcheck2/src/txnIdSelfCheck/procedures/TRURSwapTables.java]: Good idea. I'll add a base class for TRURSwapTables, TRUPSwapTablesSP & TRUPSwapTablesMP, to do that.

**Comment by johnTcrawford** [comment]
Closing this very old pull request. I believe a version of these changes was merged to master a long time ago.

**Pull request closed without merging**